src/Data/Foreign/Class.purs

```diff
@@ -56,6 +56,15 @@ instance arrayIsForeign :: (IsForeign a) => IsForeign (Array a) where
     readElement :: forall a. (IsForeign a) => Int -> Foreign -> F (Array a)
     readElement i value = readWith (ErrorAtIndex i) value
 
+instance eitherIsForeign :: (IsForeign l, IsForeign r) => IsForeign (Either l r) where
+  read value =
+    case (read value) :: F r of
+         Right rval -> pure (Right rval)
```

Comment:
Yup! My thinking is: Attempt to parse the thing we actually want, and if that fails, then attempt to parse the error case. If you have `Either String SomeType` then `String` would always succeed even if SomeType.